src/identity/components/GlobalHeader/GlobalHeaderDropdown/CreateOrganization/CreateOrganizationContext.tsx

```diff
@@ -0,0 +1,166 @@
+  changeCurrentProvider: (_: ProviderID) => null,
+  changeCurrentRegion: (_: string) => null,
+  changeOrgName: (_: string) => null,
+  clusters: {},
+  createOrg: () => null,
+  createOrgLoadingStatus: RemoteDataState.NotStarted,
+  currentRegion: '',
+  currentProvider: null,
+  orgName: '',
+}
+
+export const CreateOrgContext = createContext<CreateOrgContextType>(
+  DEFAULT_CREATE_ORG_CONTEXT
+)
+
+interface ClusterMap {
+  [providerId: string]: Cluster[]
+}
+
+export const CreateOrgProvider: FC<Props> = ({children}) => {
+  const [orgName, setOrgName] = useState(DEFAULT_CREATE_ORG_CONTEXT.orgName)
+  const [clusters, setClusters] = useState<ClusterMap>(
+    DEFAULT_CREATE_ORG_CONTEXT.clusters
+  )
+  const [currentProvider, setCurrentProvider] = useState<ProviderID>(
+    DEFAULT_CREATE_ORG_CONTEXT.currentProvider
+  )
+  const [currentRegion, setCurrentRegion] = useState<string>(
+    DEFAULT_CREATE_ORG_CONTEXT.currentRegion
+  )
+  const [createOrgLoadingStatus, setCreateOrgLoadingStatus] =
+    useState<RemoteDataState>(RemoteDataState.NotStarted)
+
+  const changeCurrentProvider = useCallback(
+    (provider: ProviderID) => {
+      setCurrentProvider(provider)
+    },
+    [setCurrentProvider]
+  )
+
+  const changeCurrentRegion = useCallback(
+    (region: string) => {
+      setCurrentRegion(region)
+    },
+    [setCurrentRegion]
+  )
+
+  const createOrg = useCallback((): Promise<void> => {
+    if (createOrgLoadingStatus === RemoteDataState.Loading) {
+      return new Promise(resolve => resolve(null))
+    }
+
+    setCreateOrgLoadingStatus(RemoteDataState.Loading)
+    return createNewOrg({
+      orgName,
+      provider: currentProvider,
+      region: currentRegion,
+    })
+      .then(() => {
+        setCreateOrgLoadingStatus(RemoteDataState.Done)
+      })
+      .catch(e => {
+        setCreateOrgLoadingStatus(RemoteDataState.Error)
+
+        throw e
+      })
+  }, [orgName, createOrgLoadingStatus, currentProvider, currentRegion])
+
+  const changeCreateOrgLoadingStatus = useCallback(
+    (status: RemoteDataState) => {
+      setCreateOrgLoadingStatus(status)
+    },
+    [setCreateOrgLoadingStatus]
+  )
+
+  useEffect(() => {
+    fetchClusterList().then(clusterData => {
+      const initialMap: ClusterMap = {}
+      const clusterMap = clusterData.reduce((previousMap, currentItem) => {
+        if (!previousMap?.[currentItem.providerId]) {
+          previousMap[currentItem.providerId] = []
+        }
+        previousMap[currentItem.providerId].push(currentItem)
+
+        return previousMap
+      }, initialMap)
+      const providers = Object.keys(clusterMap)
+      providers.forEach(providerId => {
+        clusterMap[providerId] = clusterMap[providerId].sort(
+          (a, b) => a.priority - b.priority
+        )
+      })
+      setClusters(clusterMap)
+      if (!!providers.length) {
+        setCurrentProvider(providers[0] as ProviderID)
+        setCurrentRegion(clusterMap[providers[0]][0].regionId)
```

Comment:
Looks like we have a forced type in the `as` here.

Reply:
How many properties deep is this `clusterMap`?  Since `sort()` is in-place, I think we may need to spread or deep-clone to avoid inadvertently mutating the array before we hit `setClusters()`.

Reply:
yes, `as` is a need here as I mentioned in the `AWS`, `Azure` comment. 

`setClusters` is a one time operation. This been said, it's an object with `ProviderID` as keys and array of cluster objects as its value. We don't need to do a spread here.